cirq-core/cirq/sim/simulator.py

```diff
@@ -235,6 +235,79 @@ def compute_amplitudes_sweep_iter(
+        should prefer `run()` or `simulate()`, respectively, as this method
+        only accelerates sampling for amplitude-based simulators.
+
+        Args:
+            circuit: The circuit to simulate.
+            param_resolver: Parameters to run with the program.
+            repetitions: The number of repetitions to simulate.
+            qubit_order: Determines the canonical ordering of the qubits. This
+                is often used in specifying the initial state, i.e. the
+                ordering of the computational basis states.
+            seed: Random state to use as a seed. Defaults to a random seed.
+
+        Returns:
+            A dict of bitstrings sampled from the final state of `circuit` to
+            the number of occurrences of that bitstring.
+
+        Raises:
+            ValueError: if 'circuit' has non-unitary elements, as differences
+                in behavior between sampling steps break this algorithm.
+        """
+        prng = value.parse_random_state(seed)
+        qubits = ops.QubitOrder.as_qubit_order(qubit_order).order_for(circuit.all_qubits())
+        base_circuit = circuits.Circuit(ops.I(q) for q in qubits) + circuit.unfreeze()
+        qmap = {q: i for i, q in enumerate(qubits)}
+        current_samples = {(0,) * len(qubits): repetitions}
+        solved_circuit = protocols.resolve_parameters(base_circuit, param_resolver)
+        if not protocols.has_unitary(solved_circuit):
+            raise ValueError("sample_from_amplitudes does not support non-unitary behavior.")
+        if protocols.is_measurement(solved_circuit):
+            raise ValueError("sample_from_amplitudes does not support intermediate measurement.")
+        for m_id, moment in enumerate(solved_circuit):
+            if m_id == 0:
+                continue  # skip the identities moment
+            circuit_prefix = solved_circuit[:m_id]
+            for t, op in enumerate(moment.operations):
+                new_samples: Dict[Tuple[int, ...], int] = collections.defaultdict(int)
+                qubit_indices = {qmap[q] for q in op.qubits}
+                subcircuit = circuit_prefix + ops.Moment(moment.operations[: t + 1])
+                for current_sample, count in current_samples.items():
+                    sample_set = [current_sample]
+                    for idx in qubit_indices:
+                        sample_set = [
+                            target[:idx] + (result,) + target[idx + 1 :]
+                            for target in sample_set
+                            for result in [0, 1]
+                        ]
+                    bitstrings = [int(''.join(map(str, sample)), base=2) for sample in sample_set]
+                    amps = self.compute_amplitudes(subcircuit, bitstrings, qubit_order=qubit_order)
+                    weights = [abs(a ** 2) for a in amps]
+                    norm = 1.0 / sum(weights)
+                    weights = [w * norm for w in weights]
```

Comment:
Can we convert all of these to numpy operations ?

Reply:
Done.